Homework/lesson 17/task1.py

```diff
@@ -0,0 +1,202 @@
+
+    # Group books by year
+    the_count = library.group_by_year(1846)
+    the_rape = library.group_by_year(1594)
+
+    # Print out the results
+    print(library)
+    print(Dumas)
+    print(Shakespeare)
+    print(the_count)
+    print(the_rape)
+
+
+#task3
+class Fraction:
+    def __init__(self, numerator, denominator):
+        if denominator == 0:
+            raise ZeroDivisionError("Denominator cannot be zero")
+        self.numerator = numerator
+        self.denominator = denominator
+        self.reduce()
+
+    def __str__(self):
+        return f"{self.numerator}/{self.denominator}"
+
+    def __repr__(self):
+        return f"Fraction({self.numerator}, {self.denominator})"
+
+    def __eq__(self, other):
+        if isinstance(other, Fraction):
+            self.reduce()
+            other.reduce()
+            return self.numerator == other.numerator and self.denominator == other.denominator
+        return False
+
+    def __lt__(self, other):
+        if isinstance(other, Fraction):
+            self.reduce()
+            other.reduce()
+            return self.numerator * other.denominator < other.numerator * self.denominator
+        return False
+
+    def __add__(self, other):
+        if isinstance(other, Fraction):
+            numerator = self.numerator * other.denominator + self.denominator * other.numerator
+            denominator = self.denominator * other.denominator
+            return Fraction(numerator, denominator)
+        raise TypeError(f"Cannot add {type(self).__name__} and {type(other).__name__}")
+
+    def __sub__(self, other):
+        if isinstance(other, Fraction):
+            numerator = self.numerator * other.denominator - self.denominator * other.numerator
+            denominator = self.denominator * other.denominator
+            return Fraction(numerator, denominator)
+        raise TypeError(f"Cannot subtract {type(self).__name__} and {type(other).__name__}")
+
+    def __mul__(self, other):
+        if isinstance(other, Fraction):
+            numerator = self.numerator * other.numerator
+            denominator = self.denominator * other.denominator
+            return Fraction(numerator, denominator)
+        raise TypeError(f"Cannot multiply {type(self).__name__} and {type(other).__name__}")
+
+    def __truediv__(self, other):
+        if isinstance(other, Fraction):
+            if other.numerator == 0:
+                raise ZeroDivisionError("Cannot divide by zero")
+            numerator = self.numerator * other.denominator
+            denominator = self.denominator * other.numerator
+            return Fraction(numerator, denominator)
+        raise TypeError(f"Cannot divide {type(self).__name__} and {type(other).__name__}")
+
+    def reduce(self):
+        gcd = self._gcd(self.numerator, self.denominator)
+        self.numerator //= gcd
+        self.denominator //= gcd
+
+    def _gcd(self, a, b):
+        if b == 0:
+            return a
+        return self._gcd(b, a % b)
```

Comment:
Looks too scary... :)

Reply:
it's Euclid's algorithm, I looked it up on chatGPT. it looks for a common number divisible by a and b